keps/sig-windows/20190424-windows-cri-containerd.md

```diff
@@ -151,34 +155,284 @@ Operators could deploy additional RuntimeClasses with more granular control for
+    * Labels
+      * `kubernetes.io/os` = `windows`
+      * `beta.kubernetes.io/osversion` = `Windows1809`
+1. One running `Windows Server version 1903` supporting Hyper-V isolation with `containerd`
+    * Labels
+      * `kubernetes.io/os` = `windows`
+      * `beta.kubernetes.io/osversion` = `Windows1903`
+      * `capability.microsoft.com/isolation` = `hyperv`
+
+A new RuntimeClass could be created to opt-in to `hyperv` isolation for existing workloads:
+
+```yaml
+apiVersion: node.k8s.io/v1beta1
+kind: RuntimeClass
+metadata:
+  name: windows-1809-amd64-hyperv
+handler: runhcs-wcow-hypervisor-1809  # The name of the corresponding CRI configuration
+nodeSelector:
+  kubernetes.io/os: 'windows'
+  kubernetes.io/arch: 'amd64'
+  beta.kubernetes.io/osversion: 'windows1903'
+  capability.microsoft.com/isolation: 'hyperv'
+```
+
+The new version can be added with a new RuntimeClass:
+
+```yaml
+apiVersion: node.k8s.io/v1beta1
+kind: RuntimeClass
+metadata:
+  name: windows-1903-hyperv
+handler: runhcs-wcow-hypervisor-1903  # The name of the corresponding CRI configuration
+nodeSelector:
+  kubernetes.io/os: 'windows'
+  kubernetes.io/arch: 'amd64'
+  beta.kubernetes.io/osversion: 'windows-1903'
+  capability.microsoft.com/isolation: 'hyperv'
+```
+
+Pods would need to be updated to migrate to the new nodes using `hyperv` isolation.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: mypod
+spec:
+  runtimeClassName: windows-1809-hyperv
+  # ...
+```
+
+The same approach of block with [Gatekeeper], audit existing workloads, update deployments, cordon then remove old nodes as in Option A could be followed.
+
+Once the first Hyper-V migration is done, future Windows versions could be used by just updating `RuntimeClass.NodeSelector` to use the new OS version. Cordon the old nodes, wait for the pods to move, then delete the old nodes.
+
+##### Option C: Support multiarch os/arch/version in CRI
+
+The Open Container Initiative specifications for container runtime support specifying the architecture, os, and version when pulling and starting a container. This is important for Windows because there is no kernel compatibility between major versions. In order to successfully start a container with process isolation, the right `os.version` must be pulled and run. Hyper-V can provide backwards compatibility, but the image pull and sandbox creation need to specify `os.version` because the kernel is brought up when the sandbox is created. The same challenges exist for Linux as well because multiple CPU architectures can be supported - for example armv7 with qemu and binfmt_misc.
+
+One way to make the experience uniform for dealing with multi-arch images is to add new optional fields to force a deployment to use a specific os/version/arch. This may be combined with RuntimeClass to simplify node placement if needed.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: mypod
+spec:
+  os: windows
+  osversion: 1809
+  architecture: amd64
```

Comment:
open issue - would also need variant. per OCI distribution, there can be multiple architectures with different variants. see arm/v7, arm/v8